sweep: add docs and remove dead code (#8670)
lightningnetwork/lnd: sweep/README.md

transaction - some of the outputs are time-sensitive, and may result in fund
loss if not confirmed in time. On the other hand, we don’t want to pay more
than what we can get back - if a sweeping transaction spends more than what is
meant to be swept, we are losing money due to fees.

To properly handle the case, the concept `budget` and `deadline` are introduced
to `lnd` since `v0.18.0` - for each new sweeping request, the sweeper requires
the caller to specify a deadline and a budget so it can make economic
decisions. A fee function is then created based on the budget and deadline,
which proposes a fee rate to use for the sweeping transaction. When a new block
arrives, unless the transaction is confirmed or the budget is used up, the
sweeper will perform a fee bump on it via RBF.

## Package Structure

On a high level, an UTXO is offered to the sweeper via `SweepInput`. The
sweeper keeps track of the pending inputs. When a new block arrives, it asks
the `UtxoAggregator` to group all the pending inputs into batches via
`ClusterInputs`. Each batch is an `InputSet`, and is sent to the `Bumper`. The
`Bumper` creates a `FeeFunction` and a sweeping transaction using the

> `InputSet`, and monitors its confirmation status. Every Time it's not confirmed

Author: fixed!

when a new block arrives, the `Bumper` will perform an RBF by calling
`IncreaseFeeRate` on the `FeeFunction`.

```mermaid
flowchart LR
        subgraph SweepInput
        UTXO1-->sweeper
      UTXO2-->sweeper
        UTXO3-->sweeper
        UTXO["..."]-->sweeper
        sweeper
    end

    subgraph ClusterInputs
        sweeper-->UtxoAggregator
      UtxoAggregator-->InputSet1
        UtxoAggregator-->InputSet2
        UtxoAggregator-->InputSet["..."]
    end

    subgraph Broadcast
            InputSet1-->Bumper
            InputSet2-->Bumper
            InputSet-->Bumper
    end

    subgraph IncreaseFeeRate
        FeeFunction-->Bumper
    end

        block["new block"] ==> ClusterInputs
```

#### `UtxoAggregator` and `InputSet`

`UtxoAggregator` is an interface that handles the batching of inputs.
`BudgetAggregator` implements this interface by grouping inputs with the same
deadline together. Inputs with the same deadline express the same time
sensitivity so it makes sense to sweep them in the same transaction. Once
grouped, inputs in each batch are sorted based on their budgets. The only
exception is inputs with `ExclusiveGroup` flag set, these inputs will be swept

> alone.

Author: done!

Once the batching is finished, an `InputSet` is returned, which is an interface
used to decide whether a wallet UTXO is needed or not when creating the
sweeping transaction. `BudgetInputSet` implements this interface by checking
the sum of the output values from these inputs against the sum of their
budgets - if the total budget cannot be covered, one or more wallet UTXOs are

> needed.

Author: updated!

#### `Bumper`

`Bumper` is a transaction creator, publisher, and monitor that works on an
`InputSet`. Once a sweeping transaction is created using the `InputSet`, the

> `Bumper` will monitor its confirmation status, and perform an RBF if the

Author: done!

transaction is not confirmed in the next block. It relies on the `FeeFunction`
to give the fee rate to be used.

`TxPublisher` implements the `Bumper` interface. When a transaction is created
for the first time, unless its budget has been used up, `TxPublisher` will
guarantee that the initial publish meets the RBF requirements.

#### `FeeFunction`

`FeeFunction` is an interface that specifies a function over a starting fee
rate, an ending fee rate, and a width (the deadline delta). It's used by the
`Bumper` to suggest a new fee rate for bumping the sweeping transaction.

`LinearFeeFunction` implements this interface using a linear function - it
calculates a fee rate delta using `(ending_fee_rate - starting_fee_rate) /
deadline`, and increases the fee rate by this delta value everytime a new block
arrives. Once the deadline is passed, `LinearFeeFunction` will cap its
returning fee rate at the ending fee rate.

The starting fee rate is the estimated fee rate using the deadline as the conf
target, which can be overridden by setting the `--sat_per_vbyte` via `bumpfee`
cli. The ending fee rate is the value from `budget/tx_vbyte`, which can be

> overridden by setting the `--budget` via `bumpfee` cli.

Author: updated the docs and added an example!

> overridden by setting the `--budget` via `bumpfee` cli.

Author: > Also, it's confusing that --sat_per_vbyte only sets the initial fee rate. Intuitively the user would expect the feerate to remain constant. Perhaps a rename (--initial_sat_per_vbyte) or a change of behavior is in order.

Yeah initially did that, then feel maybe we can change it in the next release, kinda wanna slowly change the RPC endpoints.

A force close transaction may have the following outputs,

- Commit outputs, which are the  `to_local` and `to_remote` outputs.
- HTLC outputs, which are the  `incoming_htlc` and `outgoing_htlc` outputs.
- Anchor outputs, which are the local and remote anchor outputs.

#### Sweeping Commit Outputs

The only output we can spend is the `to_local` output. Because it can only be
spent using our signature, there’s no time pressure here. By default, the
sweeper will use a deadline of 1008 blocks as the confirmation target for
non-time-sensitive outputs.  To overwrite the default, users can specify a
value using the config `--sweeper.nodeadlineconftarget`.

To specify the budget, users can use `--sweeper.budget.tolocal` to set the max
allowed fees in sats, or use `--sweeper.budget.tolocalratio` to set a
proportion of the `to_local` value to be used as the budget.

#### Sweeping HTLC Outputs

When facing a local force close transaction, HTLCs are spent in a two-stage
setup - the first stage is to spend the outputs using pre-signed HTLC
success/timeout transactions, the second stage is to spend the outputs from
these success/timeout transactions. All these outputs are automatically handled

> by `lnd`, in specific,

Reviewer: ```suggestion
by `lnd`. In specific,
```

Author: done!

- For an incoming HTLC in stage one, the deadline is specified using its CLTV
  from the timeout path. This output is time-sensitive.
- For an outgoing HTLC in stage one, the deadline is derived from its
  corresponding incoming HTLC’s CLTV. This output is time-sensitive.
- For both incoming and outgoing HTLCs in stage two, because they can only be
  spent by us, there is no time pressure to confirm them under a deadline.

When facing a remote force close transaction, HTLCs can be directly spent from

> this transaction, and they are both time-sensitive outputs.

Reviewer: ```suggestion
the commitment transaction., and both incoming and outgoing HTLCs are time-sensitive.
```

Author: done!

By default, `lnd` will use 50% of the HTLC value as its budget. To customize
it, users can specify `--sweeper.budget.deadlinehtlc` and
`--sweeper.budget.deadlinehtlcratio` for time-sensitive HTLCs, and
`--sweeper.budget.nodeadlinehtlc` and `--sweeper.budget.nodeadlinehtlcratio`
for non-time-sensitive sweeps.

#### Sweeping Anchor Outputs

> An anchor output is a special output that functions as “anchor” to speed up the
> unconfirmed force closing transaction via CPFP. Most of the time, anchor
> outputs won’t be swept due to it’s uneconomical. However, if the force close
> transaction contains time-sensitive outputs (HTLCs), and stays unconfirmed, the
> anchor output will be swept to accelerate the force close process.

Reviewer: > Most of the time, anchor outputs won't be swept...

Is this true?  Probably we don't need to mention frequency at all.  For example:

```suggestion
An anchor output is a special output that functions as “anchor” to speed up the
unconfirmed force closing transaction via CPFP. If the force close transaction doesn't contain any HTLCs, the anchor
output is generally uneconomical to sweep and will be ignored. However, if the force close
transaction does contain time-sensitive outputs (HTLCs), the
anchor output will be swept to CPFP the transaction and accelerate the force close process.
```

Author: thanks and updated!

>Is this true?

Yeah with the new sweeper, unless it's for CPFP, the anchor output won't be aggregated at all due to it comes with a budget (330) that cannot cover it's min relay fee.

> For CPFP-purpose anchor sweeping, its deadline is derived from the smallest
> deadline values of the HTLCs on the force close transaction. Its budget,
> however, is calculated differently - it cannot be a ratio of the anchor output
> because the value is too small (330 sats). Since its purpose is to accelerate
> the force close transaction so the time-sensitive outputs can be swept, the
> budget is in fact a ratio of what we call “value under protection”, which is
> the sum of all HTLC outputs minus sum of their budgets. By default, 50% of this
> value is used as the budget, to customize it, either use
> `--sweeper.budget.anchorcpfp` to specify sats, or use
> `--sweeper.budget.anchorcpfpratio` to specify a ratio.

Reviewer: ```suggestion
For CPFP-purpose anchor sweeping, the deadline is the closest
deadline value of all the HTLCs on the force close transaction. The budget,
however, cannot be a ratio of the anchor output
because the value is too small to contribute meaningful fees (330 sats). Since its purpose is to accelerate
the force close transaction so the time-sensitive outputs can be swept, the
budget is actually drawn from what we call “value under protection”, which is
the sum of all HTLC outputs minus the sum of their budgets. By default, 50% of this
value is used as the budget, to customize it, either use
`--sweeper.budget.anchorcpfp` to specify sats, or use
`--sweeper.budget.anchorcpfpratio` to specify a ratio.
```

Side note, I still think the behavior here is quite confusing.  Ideally the user just has to give one budget -- the deadline-aware budget, and commitment/HTLC fees would be automatically calculated.

Author: updated!

> the deadline-aware budget, and commitment/HTLC fees would be automatically calculated.

I like it - looks like I didn't get it before, we do need to figure how to allocate the budgets here tho, but I agree it makes `lnd` easier to use, probably should have an `advanced` config section.
